src/App.ts

```diff
@@ -81,26 +81,17 @@ export interface AppOptions {
 export { LogLevel, Logger } from '@slack/logger';
 
 /** Authorization function - seeds the middleware processing and listeners with an authorization context */
-export interface Authorize {
-  (source: AuthorizeSourceData | OrgAuthorizeSourceData, body?: AnyMiddlewareArgs['body']): Promise<AuthorizeResult>;
+export interface Authorize<IsEnterpriseInstall extends boolean = false> {
+  (source: AuthorizeSourceData<IsEnterpriseInstall>, body?: AnyMiddlewareArgs['body']): Promise<AuthorizeResult>;
 }
 
 /** Authorization function inputs - authenticated data about an event for the authorization function */
-export interface AuthorizeSourceData {
-  teamId: string;
-  enterpriseId?: string;
+export interface AuthorizeSourceData<IsEnterpriseInstall extends boolean = false> {
+  teamId: IsEnterpriseInstall extends true ? string | undefined : string;
+  enterpriseId: IsEnterpriseInstall extends true ? string : string | undefined;
   userId?: string;
   conversationId?: string;
-  isEnterpriseInstall?: boolean;
-}
-
-/** Authorization function inputs - authenticated data about an event for the authorization function */
-export interface OrgAuthorizeSourceData {
-  enterpriseId: string;
-  teamId?: string;
-  userId?: string;
-  conversationId?: string;
-  isEnterpriseInstall?: boolean;
+  isEnterpriseInstall: IsEnterpriseInstall;
 }
```

Comment:
This is likely the most meaningful refactor. I've done something similar here to what I did in `node-slack-sdk` - consolidated the types that were very similar but just varied on `IsEnterpriseInstall` by making the types generic.

This allowed me to safely reduce code repetition within `App.processEvent()` and `singleTeamAuthorization()` (now `singleAuthorization()`).